File: Leetcode/cpp/sol543.cpp

// https://leetcode.com/problems/diameter-of-binary-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int longest;

    Solution(){
        longest = 0;
    }

    int helper(TreeNode* root){
        if(!root){
            return 0;
        }
        int left = helper(root->left);
        int right = helper(root->right);
        int diameter = left + right;
        longest = max(longest, diameter);
        return 1 + max(left, right);
    }

    int diameterOfBinaryTree(TreeNode* root) {
        helper(root);
        return longest;
    }
};

File: Leetcode/typescript/sol543.ts

//https://leetcode.com/problems/diameter-of-binary-tree/


/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function diameterOfBinaryTree(root: TreeNode | null): number {
    let result = 0;
    const helper = (node:TreeNode | null):number => {
        if(!node){
            return 0;
        }
        let left = helper(node.left);
        let right =  helper(node.right)
        result = Math.max(result, left+right)
        return 1 + Math.max(left, right)
    }

    return result
};